### ISSUE TITLE
How do I decode beta version code?

### ISSUE THREAD

**Issue by harryqt** [xy2iii/BlockAdBlock-reversed#1]
I understand that replacing `eval()` with `console.log()` with the basic script will output the decoded codes. However the beta-version does not use `eval()`; How do I decode that?

**Comment by xy2i** [comment]
Hi, thanks for the heads up. I totally missed that it was around.
I made an explanation of this, but just skip to the end if you're not interested and want the source.

After prettifying we get something like this:
https://gist.github.com/xy2iii/0f4920d8c815556172dd031a714bf7bb

On [line 11](https://gist.github.com/xy2iii/0f4920d8c815556172dd031a714bf7bb#file-blockadblock-beta-version-420b-L14) we can find what appears to be the payload, because it looks a lot like Dean's Edward packer code.
And on [line 31](https://gist.github.com/xy2iii/0f4920d8c815556172dd031a714bf7bb#file-blockadblock-beta-version-420b-L31) there's some [JSFuck](https://en.wikipedia.org/wiki/JSFuck) which may do something.

And honestly, that looks like a lot of complicated code. Maybe we can short-circuit it? At line 30 there's an interesting thing: what looks to be the payload (which started as an empty string) is [assigned to x](https://gist.github.com/xy2iii/0f4920d8c815556172dd031a714bf7bb#file-blockadblock-beta-version-420b-L30).

Since x sticks around, we can paste the script and try to run it. After letting it run, then looking at x in the console, we get [the entire payload deobfuscated](https://gist.github.com/xy2iii/31fb33c946510c08e2c44281c4f387e5).

Still, what does the last line, the JSFuck do? A JSFuck fragment is just an expression, so it can only do anything that an expression would do. If I were to obfuscate something, I'd make an [immediately invoked function expression](https://developer.mozilla.org/en-US/docs/Glossary/IIFE), so that way the expression could declare a function which could do anything, and then immediately execute it, by calling it.

JSFuck still follows all the rules of JavaScript: to invoke a function, we need to put parantheses at the end. What if we try to execute the JSFuck fragment in the console, but omit the suspicious parantheses at the end? The console tells us that a "function anonymous()" has been declared.

In JS, you can obtain the source code of objects (and functions) by calling [toSource](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/toSource) on them. And if we remplace the parantheses by a toSource() call, we get the unobfuscated source of the JSFuck fragment in the console:
```js
(function anonymous(
) {
eval(x)
})
```

This declares an anonymous function that evals whatever x variable is set.

To recap:
- The code does some new obfuscation, that I didn't bother reversing.
- It stores the unobfuscated, packed code in a variable called x.
- It then calls a JSFuck snipped, which creates an IIFE that evals x.
- Afterwards, Dean Edward's packer unpacks the code and it executes.

Here's the [unobfuscated source](https://gist.github.com/xy2iii/31fb33c946510c08e2c44281c4f387e5).
Thank you for letting me know about this! :smile:
